lib/endpoint.js

```diff
@@ -90,17 +90,89 @@ VertoEndpoint.prototype.attemptInvite = function(vertoCall, matrixSide, force) {
+    // and sum the component IDs for each foundation (which should sum to 1 or 3
+    // depending on whether we've received just RTP or RTP+RTCP)
+    var stats = [];
+
+    // create empty stats blocks where we expect to see candidates so we can
+    // spot if candidates are entirely missing...
+    for (var i = 0; i < matrixSide.offer.match(/^m=/mg).length; i++) {
+        stats[i] = {};
+    }
+
     for (var i = 0; i < matrixSide.candidates.length; i++) {
         var c = matrixSide.candidates[i];
         if (!c.candidate) { continue; }
-        // got enough candidates when SDP has a srflx or relay candidate
-        if (c.candidate.indexOf("typ srflx") !== -1 ||
-                c.candidate.indexOf("typ relay") !== -1) {
-            enoughCandidates = true;
-            console.log("Gathered enough candidates for %s", matrixSide.mxCallId);
-            break; // bail early
+
+        var mLineIndex;
+        if (c.sdpMLineIndex !== undefined) {
+            mLineIndex = c.sdpMLineIndex;
+        }
+        else {
+            // hack: synthesise an mLineIndex based on heuristics
+            if (c.sdpMid === 'audio') {
+                mLineIndex = 0;
+            }
+            else if (c.sdpMid === 'video') {
+                mLineIndex = 1;
+            }
+            else {
+                mLineIndex = 0; // this is probably wrong though.
+            }
+        }
+        stats[mLineIndex] = stats[mLineIndex] || {};
+        if (c.candidate.indexOf("typ host") !== -1) {
+            stats[mLineIndex].hasHost = 1;
+        }
+        if (c.candidate.indexOf("typ srflx") !== -1) {
+            stats[mLineIndex].hasSrflx = 1;
+        }
+        if (c.candidate.indexOf("typ relay") !== -1) {
+            stats[mLineIndex].hasRelay = 1;   
         }
+        var m = c.candidate.match(/^candidate:(.*?) (.*?) /);
+        if (m) {
+            stats[mLineIndex].componentIdSumByFoundation =
+                stats[mLineIndex].componentIdSumByFoundation || {};
+            stats[mLineIndex].componentIdSumByFoundation[m[1]] += m[2];
+        }
+        else {
+            console.error("Can't parse candidate: " + c.candidate);
+        }
+    }
+
+    var enoughCandidates = true; // start off optimistic...
+    stats.forEach(function(stat) {
+        // we've gathered enough candidates when've received srflx or relay candidates
+        //
+        // this works because if we've got a srflx, we will always be able to use it
+        // given we know that the freeswitch isn't NATted - so if a NATted client
+        // is able to correctly discover a srflx candidate, it should also be able to
+        // talk through to FS.
+        //
+        // Similarly, if we've seen a relay candidate (which are typically the slowest
+        // to set up and last to arrive), we know that will always work too, so we don't
+        // need to wait for any others.
+
+        if (!stat.hasHost) enoughCandidates = false;
```

Comment:
I hearts the prose above, but then this is the first thing you do. You don't explain why not having a host means we don't have enough candidates.